extensions/rocksdb-repos/controllers/RocksDbPersistableKeyValueStoreService.cpp

```diff
@@ -0,0 +1,200 @@
+
+#include "utils/StringUtils.h"
+
+#include <fstream>
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace controllers {
+
+core::Property RocksDbPersistableKeyValueStoreService::Directory(
+    core::PropertyBuilder::createProperty("Directory")->withDescription("Path to a directory for the database")
+        ->isRequired(true)->build());
+
+RocksDbPersistableKeyValueStoreService::RocksDbPersistableKeyValueStoreService(const std::string& name, const std::string& id)
+    : controllers::KeyValueStoreService(name, id)
+    , controllers::AbstractAutoPersistingKeyValueStoreService(name, id)
+    , db_valid_(false)
+    , logger_(logging::LoggerFactory<RocksDbPersistableKeyValueStoreService>::getLogger()) {
+}
+
+RocksDbPersistableKeyValueStoreService::RocksDbPersistableKeyValueStoreService(const std::string& name, utils::Identifier uuid /*= utils::Identifier()*/)
+    : controllers::KeyValueStoreService(name, uuid)
+    , controllers::AbstractAutoPersistingKeyValueStoreService(name, uuid)
+    , db_valid_(false)
+    , logger_(logging::LoggerFactory<RocksDbPersistableKeyValueStoreService>::getLogger()) {
+}
+
+RocksDbPersistableKeyValueStoreService::~RocksDbPersistableKeyValueStoreService() {
+  if (db_valid_) {
+    delete db_;
+  }
+}
+
+void RocksDbPersistableKeyValueStoreService::initialize() {
+  AbstractAutoPersistingKeyValueStoreService::initialize();
+  std::set<core::Property> supportedProperties;
+  supportedProperties.insert(Directory);
+  updateSupportedProperties(supportedProperties);
+}
+
+void RocksDbPersistableKeyValueStoreService::onEnable() {
+  if (configuration_ == nullptr) {
+    logger_->log_debug("Cannot enable RocksDbPersistableKeyValueStoreService");
+    return;
+  }
+
+  AbstractAutoPersistingKeyValueStoreService::onEnable();
+
+  if (!getProperty(Directory.getName(), directory_)) {
+    logger_->log_error("Invalid or missing property: Directory");
+    return;
+  }
+
+  if (db_valid_) {
+    db_valid_ = false;
+    delete db_;
+  }
+  rocksdb::Options options;
+  options.create_if_missing = true;
+  if (!always_persist_) {
+    options.manual_wal_flush = true;
+  }
+  rocksdb::Status status = rocksdb::DB::Open(options, directory_, &db_);
+  if (status.ok()) {
+    db_valid_ = true;
+    logger_->log_trace("Successfully opened RocksDB database at %s", directory_.c_str());
+  } else {
+    logger_->log_error("Failed to open RocksDB database at %s, error: %s", directory_.c_str(), status.getState());
+    return;
+  }
+
+  if (always_persist_) {
+    default_write_options.sync = true;
+  }
+
+  logger_->log_trace("Enabled RocksDbPersistableKeyValueStoreService");
+}
+
+void RocksDbPersistableKeyValueStoreService::notifyStop() {
+  AbstractAutoPersistingKeyValueStoreService::notifyStop();
+
+  if (db_valid_) {
+    delete db_;
+    db_valid_= false;
+  }
+}
+
+bool RocksDbPersistableKeyValueStoreService::set(const std::string& key, const std::string& value) {
+  if (!db_valid_) {
```

Comment:
Wonder if we could utilize the nullcheck macro here (used id nanofi)

Reply:
Replaced with unique_ptr, so no longer relevant.